{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mitochondria retrograde ODE model\n",
    "\n",
    "![GitHub repo size](https://img.shields.io/github/repo-size/ntumitolab/mitochondria-retrograde-model)\n",
    "\n",
    "> Codes, data and notebooks for the manuscript titled \"Mathematical Modeling and Analysis of Mitochondrial Retrograde Signaling Dynamics\"\n",
    "\n",
    "## Graphical abstract\n",
    "\n",
    "![Graphical abstract](https://user-images.githubusercontent.com/29009898/130342513-081f4592-3cc6-4468-ba3b-868416f3be6b.png)\n",
    "\n",
    "## Mitochondrial Retrograde Signalling\n",
    "\n",
    "Mitochondrial retrograde signaling reports mitochondrial status to the nucleus. However, there is a lack of understanding of how the nucleus capture mitochondrial status in dynamics and information processing. It is a complicated biochemical reaction that occurs in most eukaryotic organisms. In this repository, we focus on the RTG pathway in yeast. This pathway is the simplest retrograde signaling pathway that has been investigated thoroughly. Data are collected from [1] and [2] (See [Dataset](#dataset)). This repository aims to compose known protein interactions and nucleus relocation that fulfills all known responses of the yeast RTG pathway. Monte-Carlo approach is used to solve this Boolean satisfiability problem, and the parameter searching/ simulation/ threading is facilitated by DifferentialEquations.jl [3].\n",
    "\n",
    "## Installation\n",
    "\n",
    "Julia v1.7+ is needed (https://julialang.org/).\n",
    "\n",
    "## RTG protein levels\n",
    "\n",
    "<https://github.com/ntumitolab/rnaseq_rtg_expression>\n",
    "\n",
    "## Parameters of retrograde signaling model\n",
    "\n",
    "> Table of parameters: [`solution_rtgM4.csv`](https://github.com/ntumitolab/mitochondria-retrograde-model/blob/main/src/data/solution_rtgM4.csv)\n",
    "\n",
    "This file is generated and modified from [a script](https://github.com/stevengogogo/RetroSignalModel.jl/blob/82cc8fd91b1fa03983c08a5eed1144bac3e93dcc/scripts/find_valid_solutions.jl). All solutions are corresponding to the knockout experiments of [1] and [2] with the conditions in [boolean_table_RTG13.csv](https://github.com/ntumitolab/mitochondria-retrograde-model/blob/main/src/data/boolean_table_RTG13.csv). The solutions are stored in [`solution_rtgM4.csv`](https://github.com/ntumitolab/mitochondria-retrograde-model/blob/main/src/data/solution_rtgM4.csv).\n",
    "\n",
    "## Responses of Yeast RTG proteins to mitochondrial damage\n",
    "\n",
    "> Data: [boolean_table_RTG13.csv](https://github.com/ntumitolab/mitochondria-retrograde-model/blob/main/src/data/boolean_table_RTG13.csv)\n",
    "\n",
    "This folder contains summarized responses of mitochondrial retrograde signaling in yeast.\n",
    "\n",
    "### Components\n",
    "\n",
    "| Standard Name | Variable Name | Details                                      |\n",
    "| ------------- | ------------- | -------------------------------------------- |\n",
    "| RTG1          | `rtg1`        | https://www.yeastgenome.org/locus/S000005428 |\n",
    "| RTG2          | `rtg2`        | https://www.yeastgenome.org/locus/S000005428 |\n",
    "| RTG3          | `rtg3`        | https://www.yeastgenome.org/locus/S000000199 |\n",
    "| Mks1          | `mks1`        | https://www.yeastgenome.org/locus/S000005020 |\n",
    "\n",
    "### Definition of Response\n",
    "\n",
    "In [1] and [2], RTG response is observed via GFP tags on either RTG1 or RTG3. In wild-type, mitochondrial damage can cause these proteins to accumulate in the nucleus, resulting in the intensified brightness of the nucleus region observed by fluorescent microscopy. As shown in [boolean_table_RTG13.csv](https://github.com/ntumitolab/mitochondria-retrograde-model/blob/main/src/data/boolean_table_RTG13.csv), the responses are categorized in binary results: whether GFP is accumulated in the nucleus in a given condition. Based on [1] and [2], there are 20 reactions listed in the table.\n",
    "\n",
    "For example, the following is one of the conditions mentioned in [1]:\n",
    "\n",
    "| Rtg1 | Rtg2 | Rtg3 | s   | Mks | gfp  | Trans2Nuc |\n",
    "| ---- | ---- | ---- | --- | --- | ---- | --------- |\n",
    "| 0    | 0    | 1    | 1   | 1   | rtg3 | 1         |\n",
    "\n",
    "Under the columns of `Rtg1`, `Rtg2`, `Rtg3` and `Mks`, `0` means that the given protein is suppressed by knockout. On the other hand, `1` represent an expression of wild type. Also, `1` in `s` represent mitochondrial dysfunction, and `0` means the absence of mitochondrial damage. The `gfp` column describes the location of GFP tag. In this example, GFP tag is on `Rtg3`. As known in [1], `Rtg3-GFP` translocates to the nucleus under this condition. Therefore, `Trans2Nuc` is marked as `1`, which means the GFP tags nucleus translocation happens.\n",
    "\n",
    "### Reactions\n",
    "\n",
    "There are 20 reactions summarized in the table. Some conditions are yet to be explored; some are from [1] (Sekito et al. 2000) or [2] (Sekito et al. 2002). Missing conditions are labeled with `NA`.\n",
    "\n",
    "| Line Number | Reference |\n",
    "| ----------- | --------- |\n",
    "| 2           | NA        |\n",
    "| 3           | [1]       |\n",
    "| 4           | [1]       |\n",
    "| 5           | [1]       |\n",
    "| 6           | [1]       |\n",
    "| 7           | [1]       |\n",
    "| 8           | [1]       |\n",
    "| 9           | [1]       |\n",
    "| 10          | NA        |\n",
    "| 11          | NA        |\n",
    "| 12          | [1]       |\n",
    "| 13          | [1]       |\n",
    "| 14          | [1]       |\n",
    "| 15          | [1]       |\n",
    "| 16          | [1]       |\n",
    "| 17          | [1]       |\n",
    "| 18          | [2]       |\n",
    "| 19          | [2]       |\n",
    "| 20          | [2]       |\n",
    "| 21          | [2]       |\n",
    "\n",
    "## References\n",
    "\n",
    "1. Sekito, Takayuki, Janet Thornton, and Ronald A. Butow. \"Mitochondria-to-nuclear signaling is regulated by the subcellular localization of the transcription factors Rtg1p and Rtg3p.\" Molecular biology of the cell 11.6 (2000): 2103-2115. URL: https://doi.org/10.1091/mbc.11.6.2103\n",
    "2. Sekito, Takayuki, Zhengchang Liu, Janet Thornton, and Ronald A. Butow. “RTG-Dependent Mitochondria-to-Nucleus Signaling Is Regulated by MKS1 and Is Linked to Formation of Yeast Prion [URE3].” Molecular Biology of the Cell 13, no. 3 (March 2002): 795–804. https://doi.org/10.1091/mbc.01-09-0473.\n",
    "3. Rackauckas, Christopher, and Qing Nie. “DifferentialEquations.Jl – A Performant and Feature-Rich Ecosystem for Solving Differential Equations in Julia.” Journal of Open Research Software 5, no. 1 (May 25, 2017): 15. https://doi.org/10.5334/jors.151.\n",
    "4. Gasch, Audrey P., et al. \"Single-cell RNA sequencing reveals intrinsic and extrinsic regulatory heterogeneity in yeast responding to stress.\" PLoS biology 15.12 (2017): e2004050. URL: https://doi.org/10.1371/journal.pbio.2004050\n",
    "5. Delmans, Mihails, and Martin Hemberg. \"Discrete distributional differential expression (D3E)-a tool for gene expression analysis of single-cell RNA-seq data.\" BMC bioinformatics 17.1 (2016): 1-13. URL: https://doi.org/10.1186/s12859-016-0944-6\n",
    "\n",
    "## Runtime information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "versioninfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Pkg\n",
    "Pkg.status()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.8.0-rc3",
   "language": "julia",
   "name": "julia-1.8"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.8.0-rc3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
